### ISSUE TITLE
Lesson 7: Resnets from scratch; U-net; Generative (adversarial) networks

### ISSUE THREAD

**Issue by datalass1** [datalass1/fastai#37]
Overview
In the final lesson of Practical Deep Learning for Coders we'll study one of the most important techniques in modern architectures: the skip connection. This is most famously used in the resnet, which is the architecture we've used throughout this course for image classification, and appears in many cutting edge results. We'll also look at the U-net architecture, which uses a different type of skip connection to greatly improve segmentation results (and also for similar tasks where the output structure is similar to the input).

We'll then use the U-net architecture to train a super-resolution model. This is a model which can increase the resolution of a low-quality image. Our model won't only increase resolution—it will also remove jpeg artifacts, and remove unwanted text watermarks.

In order to make our model produce high quality results, we will need to create a custom loss function which incorporates feature loss (also known as perceptual loss), along with gram loss. These techniques can be used for many other types of image generation task, such as image colorization.

Finally, we'll learn about a recent loss function known as generative adversarial loss (used in generative adversarial networks, or GANs), which can improve the quality of generative models in some contexts, at the cost of speed.

The techniques we show in this lesson include some unpublished research that:

Let us train GANs more quickly and reliably than standard approaches, by leveraging transfer learning
Combines architectural innovations and loss function approaches that haven't been used in this way before.
The results are stunning, and train in just a couple of hours (compared to previous approaches that take a couple of days).

**Comment by datalass1** [comment]
# Building a Deep Learning Architecture from scratch with MNIST using resnet
Our image item list contains 70,000 items, and it's a bunch of images that are 1 by 28 by 28.

Once you've got an image item list, you then split it into training versus validation.

Next thing we can do is to add transforms; for small images of digits like this, you just add a bit of random padding. 

Now we've got a transformed labeled list, we can pick a batch size and choose data bunch.

We can choose normalize. In this case, we're not using a pre-trained model, so there's no reason to use ImageNet stats here. So if you call normalize like this without passing in stats.

### Basic CNN with batch norm
Normalization is a process where given a set of data, you subtract from each element the mean value for that data set and divide it by the data set's standard deviation. By doing so, we put the input values onto the same "scale".

Often times with images, we don't worry about dividing by the standard deviation, but just subtract the mean.

Let's start out creating a simple CNN. The input is 28 by 28.

All of my convolution is going to be kernel size 3 stride 2 padding 1.
`def conv(ni,nf): return nn.Conv2d(ni, nf, kernel_size=3, stride=2, padding=1)`

Each time you have a convolution, it's skipping over one pixel so it's jumping two steps each time. That means that each time we have a convolution, it's going to halve the grid size. 

```
model = nn.Sequential(
    conv(1, 8), # 14
    nn.BatchNorm2d(8),
    nn.ReLU(),
    conv(8, 16), # 7
    nn.BatchNorm2d(16),
    nn.ReLU(),
    conv(16, 32), # 4
    nn.BatchNorm2d(32),
    nn.ReLU(),
    conv(32, 16), # 2
    nn.BatchNorm2d(16),
    nn.ReLU(),
    conv(16, 10), # 1
    nn.BatchNorm2d(10),
    Flatten()     # remove (1,1) grid
)
```



![resblock](https://github.com/hiromis/notes/blob/master/lesson7/8.png?raw=true)
